RavenDB-19358: Create a backup history page

ArieSLV opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19358/Create-a-backup-history-page

### Additional description

With this new feature, you can now easily keep track of all backups performed within your RavenDB cluster. You can access information about each backup through two different endpoints.

1. Backup History
To see a list of all backups made, use the "**/databases/*/backup-history**" endpoint. You can filter the information by using the following keys:
     - "taskId" key: filters information only related to backups made within a specific TaskID periodic backup.
     - "nodeTag" key: filters information only related to backups made within a specific node name.

   This endpoint will show you all backups made with the following information for each backup:
   - Type of backup;
   - When the backup was made;
   - Name of the backuped database;
   - How long it took;
   - Whether there was an error;
   - Was the backup full or incremental;
   - On which node the backup was made;
   - Time of the last full backup.;

Each entry in the backup history includes a Full Backup and all its subsequent incremental backups. The number of stored full backup entries is set to 30 by default but can be configured in the database settings.

2. Backup History Details:
The "**/databases/*/backup-history-details**" endpoint offers additional information on each backup. It provides information on the number of processed, skipped, and errored items such as Database Record, Documents, Attachments, Counters, TimeSeries, Revisions, Conflicts, Indexes, Identities, Compare Exchange, and Subscriptions, along with a complete log of the backup operation
Note that backup history details are only available when requested from a specific node in relation to a specific backup made on that node.

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Is it platform specific issue?

- No

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove that the feature works;

### Testing by RavenDB QA team

- This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). The issue on YouTrack was marked with `QA Required` tag.

### Is there any existing behavior change of other features due to this change?

- Yes: Notifications for successful backups have been turned off in the notification center because the information is now displayed in a new section.

### UI work

- It requires further work in the Studio

ArieSLV committed:
RavenDB-19358: Create a backup history page

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+    }
+
+    private void StoreInternal(string key, DynamicJsonValue value, RavenTransaction transaction = null, TransactionOperationContext context = null)
+    {
```
better to have 2 different overloads

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+        using (var scope = new DisposableScope())
+        {
+            scope.EnsureDispose(_contextPool.AllocateOperationContext(out TransactionOperationContext context));
+            scope.EnsureDispose(context.OpenReadTransaction());
```
this looks strange..

ArieSLV replied:
Removed.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+            {
+                var entry = GetCurrentHistoryEntry(context, result.Value);
+
+                Delete(entry.Key, tx);
```
if you are deleting from the `BackupHistorySchema.BackupHistory` you need to re-iterate it again.
Consider maybe using `DeleteByPrimaryKeyPrefix`

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+        var doc = new BlittableJsonReaderObject(ptr, size, context);
+        var key = Encoding.UTF8.GetString(result.Reader.Read(0, out size), size);
+
+        Transaction.DebugDisposeReaderAfterTransaction(context.Transaction.InnerTransaction, doc);
```
do we need to dispose the `doc`?

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+        return (key, JsonDeserializationClient.BackupHistoryEntry(doc));
+    }
+
+    public bool Delete(string id, RavenTransaction existingTransaction = null)
```
better to have 2 overloads here

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+
+            details = GetBackupHistoryDetails(key, context, tx);
+
+            return scope.Delay();
```
saw that in other places, but I really don't understand why we need those :)

ArieSLV replied:
Removed.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+        }
+    }
+
+    private BlittableJsonReaderObject GetBackupHistoryDetails(string key, JsonOperationContext context, RavenTransaction tx)
```
I think `TransactionOperationContext` should be enough here instead of the `JsonOperationContext` & `RavenTransaction`

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,304 @@
+
+        public static class BackupHistoryTable
+        {
+            public const int ItemKey = 0;
```
please add a comment here to describe how the key is look like
e.g. `/databases/<name>/backups/... `

ArieSLV replied:
Done.

karmeli87 commented on src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs:
```diff
@@ -880,7 +868,10 @@ public static void SaveBackupStatus(PeriodicBackupStatus status, DocumentDatabas
+                documentDatabase.ConfigurationStorage.BackupHistoryStorage.StoreBackupHistoryEntries(command);
             }
+
+            documentDatabase.ConfigurationStorage.BackupHistoryStorage.StoreBackupDetails(backupResult, status);
```
would it make sense to unify this with `RetractTemporaryStoredBackupHistoryEntries` on top in a single write-tx ?

ArieSLV replied:
I don’t see much point in this since we save the `Details` in local storage anyway, but  `RetractTemporaryStoredBackupHistoryEntries` using only in emergency cases when the cluster is down.

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -1377,7 +1381,7 @@ private static bool ValidatePropertyExistence(BlittableJsonReaderObject cmd, str
         }
 
-        private void SetValueForTypedDatabaseCommand(ClusterOperationContext context, string type, BlittableJsonReaderObject cmd, long index, Leader leader, out object result)
+        private UpdateValueForDatabaseCommand SetValueForTypedDatabaseCommand(ClusterOperationContext context, string type, BlittableJsonReaderObject cmd, long index, Leader leader, out object result, bool notifyDatabase = true)
```
~where is this used?~
when we don't want to nofify?

ArieSLV replied:
All that was done within the `UpdateBackupHistory` method was moved to the `UpdatePeriodicBackupStatusCommand.Execute` method. Now we notify only one time, as it always was.

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -4457,6 +4464,165 @@ public unsafe bool IsReplicationCertificateByPublicKeyPinningHash(TransactionOpe
             return false;
         }
+
+        private unsafe void UpdateBackupHistory(ClusterOperationContext context, string type, long index, UpdatePeriodicBackupStatusCommand command,
```
let's create a partial `ClusterStateMachine` and name it `ClusterStateMachine.Backup.cs` and move all backup related stuff there :)

ArieSLV replied:
Same here: lost relevance due to changes.

karmeli87 commented on src/Raven.Server/ServerWide/ClusterStateMachine.cs:
```diff
@@ -4457,6 +4464,165 @@ public unsafe bool IsReplicationCertificateByPublicKeyPinningHash(TransactionOpe
+
+            NotifyDatabaseAboutChanged(context, command.DatabaseName, index, type, DatabasesLandlord.ClusterDatabaseChangeType.ValueChanged, backupDetailsIdsToDelete);
+
+            void StoreEntry(BackupHistoryEntry backupHistoryEntryToAdd)
```
let's avoid having large local functions.

ArieSLV replied:
Same here: lost relevance due to changes.

karmeli87 commented on src/Raven.Server/Web/System/BackupDatabaseHandler.cs:
```diff
@@ -91,6 +95,227 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+            var nodeTagFromRequest = GetStringValuesQueryString("nodeTag", required: false);
+
+            var currentNodeTag = ServerStore.NodeTag;
+            var documentDatabase = await ServerStore.DatabasesLandlord.TryGetOrCreateResourceStore(name);
```
This looks wrong. 
If you need to access a specific database the EP should be `/databases/*/...` and inherit from the `DatabaseRequestHandler`

karmeli87 replied:
I would expect this EP to fetch a cluster-wide info

ArieSLV replied:
Created a separate Handler for BackupHistory.

ArieSLV committed:
RavenDB-19358: PR comments

ArieSLV committed:
RavenDB-19358: Minor test refactoring to improve stability

ArieSLV committed:
RavenDB-19358: Increased timeout for RunBackup tasks

ArieSLV committed:
RavenDB-19358: Added an `Id` to each `BackupHistoryEntry` within the `/backup-history/` endpoint

ml054 committed:
RavenDB-19358 Create a backup history page - studio work

ArieSLV committed:
RavenDB-19358:
 - The response now includes `TaskName` and calculated `TaskId`.
 - The `/databases/*/backup-history` endpoint has been entirely refactored.

ArieSLV committed:
RavenDB-19358: Rebase

ml054 commented on src/Raven.Studio/typings/_studio/dto.d.ts:
```diff
@@ -870,6 +870,14 @@ interface geoPointInfo extends Raven.Server.Documents.Indexes.Spatial.Coordinate
 }
 
+interface BackupHistoryItem extends Raven.Client.Documents.Operations.Backups.BackupHistoryEntry {
+    Id: string;
```
Add line:
```
TaskName: string;
```

ArieSLV replied:
Done.

ArieSLV committed:
RavenDB-19358: Studio fix

ml054 commented on src/Raven.Studio/typescript/viewmodels/database/status/backupHistory.ts:
```diff
@@ -0,0 +1,179 @@
+    }
+    
+    private getLinkForTaskView(item: BackupHistoryItem) {
+        if (item.NodeTag !== clusterTopologyManager.default.localNodeTag()) {
+            return null;
+        }
```
link should work on all nodes - we need to remove those 3 lines.

ml054 replied:
@ArieSLV I fix that. Will push shortly.

ml054 committed:
RavenDB-19358 Create a backup history page - link fix

ml054 committed:
RavenDB-19358 Create a backup history page

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupHistoryEntry.cs:
```diff
@@ -0,0 +1,76 @@
+
+public class BackupHistoryEntry : IDynamicJsonValueConvertible
+{
+    public string GenerateItemKey() => $"values/{DatabaseName}/backup-history/{TaskId}";
```
internal?

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupHistoryEntry.cs:
```diff
@@ -0,0 +1,76 @@
+public class BackupHistoryEntry : IDynamicJsonValueConvertible
+{
+    public string GenerateItemKey() => $"values/{DatabaseName}/backup-history/{TaskId}";
+    public static string GenerateItemPrefix(string databaseName) => $"values/{databaseName}/backup-history/";
```
internal

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupHistoryEntry.cs:
```diff
@@ -0,0 +1,76 @@
+    public string DatabaseName { get; set; }
+    public long? DurationInMs { get; set; }
+    public string Error { get; set; }
+    public bool IsFull { get; set; }
```
use enum BackupKind (need to move it from server probably?)

ppekrol commented on src/Raven.Client/Documents/Operations/Backups/BackupHistoryEntry.cs:
```diff
@@ -0,0 +1,76 @@
+    }
+}
+
+public class BackupHistory
```
public class in client api that looks internal, should be moved to server?

ppekrol commented on src/Raven.Server/Documents/Operations/Operations.cs:
```diff
@@ -227,6 +226,13 @@ private void RaiseNotifications(OperationStatusChange change, Operation operatio
             operation.NotifyCenter(operationChanged, x => _notificationCenter.Add(x));
 
+            // All information regarding successful backups is moved to a dedicated section, so, we are disabling all notifications for successful backups. 
+            if (operationChanged.TaskType == OperationType.DatabaseBackup && change.State.Status == OperationStatus.Completed)
```
do not think this is necessary? why we want that?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,234 @@
+
+    protected readonly Logger Logger;
+
+    private readonly TableSchema _backupHistorySchema = new();
```
this is static schema and can be initialized in static ctor?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,234 @@
+                bool deleteResult;
+                var table = transaction.InnerTransaction.OpenTable(_backupHistorySchema, BackupHistorySchema.BackupHistory);
+
+                using (Slice.From(transaction.InnerTransaction.Allocator, id, out Slice alertSlice))
```
alertSlice? is this naming correct?

ppekrol commented on src/Raven.Server/Web/System/BackupHistoryHandler.cs:
```diff
@@ -0,0 +1,64 @@
+
+internal class BackupHistoryHandler : DatabaseRequestHandler
+{
+    [RavenAction("/databases/*/backup-history-details", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
DatabaseAdmin

ppekrol replied:
`/databases/*/admin/backup/history/details`

ppekrol commented on src/Raven.Server/Web/System/BackupHistoryHandler.cs:
```diff
@@ -0,0 +1,64 @@
+        }
+    }
+
+    [RavenAction("/databases/*/backup-history", "GET", AuthorizationStatus.ValidUser, EndpointType.Read)]
```
DatabaseAdmin

ppekrol replied:
`/databases/*/admin/backup/history`

ppekrol commented on src/Raven.Server/Web/System/BackupHistoryHandler.cs:
```diff
@@ -0,0 +1,64 @@
+        using (Database.ConfigurationStorage.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+        await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBodyStream()))
+        {
+            var key = $"values/{Database.Name}/backup-history/{BackupHistoryItemType.Details}/{taskId}/{ServerStore.NodeTag}/{createdAtTicksAsId}";
```
this key should be generated via some static method used by all places where such a key is created

ppekrol commented on src/Raven.Server/Web/System/BackupHistoryHandler.cs:
```diff
@@ -0,0 +1,64 @@
+            writer.WriteStartObject();
+            writer.WritePropertyName(nameof(BackupHistoryItemType.Details));
+
+            var details = Database.ConfigurationStorage.BackupHistoryStorage.GetBackupHistoryDetails(key, context);
```
you should provide taskId and id as parameters here, what is the storage key should be handled internally

ppekrol commented on src/Raven.Server/Documents/DatabasesLandlord.cs:
```diff
@@ -1113,8 +1114,9 @@ private void NextScheduledActivityCallback(string databaseName, IdleDatabaseActi
-                            BackupUtils.SaveBackupStatus(backupStatus, databaseName, _serverStore, _logger, backupResult);
+                            
+                            var database = TryGetOrCreateResourceStore(databaseName).Result;
+                            BackupUtils.SaveBackupStatus(backupStatus, database, _logger, backupResult);
```
this method inside already uses async helpers, leave previous argumnents and just take the instance of database from the dtabaseslandlord internally

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,234 @@
+        }
+    }
+    
+    public Dictionary<string, BlittableJsonReaderObject> ReadItems(TransactionOperationContext context, BackupHistoryItemType type)
```
why do we want to deal with blittables outside of this storage?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupHistory/BackupHistoryStorage.cs:
```diff
@@ -0,0 +1,234 @@
+        return temporaryStoredEntries;
+    }
+
+    public BlittableJsonReaderObject GetBackupHistoryDetails(string key, TransactionOperationContext context)
```
why not deserialize it?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs:
```diff
@@ -58,6 +57,8 @@ public class BackupTask
         private readonly string _taskName;
         internal PeriodicBackupRunner.TestingStuff _forTestingPurposes;
         private readonly DateTime _startTimeUtc;
+        private static int MaxNumberOfFullBackupsInBackupHistory;
```
this cannot be static, can be changed per database, also even if per server then in tests you have different instances o server

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs:
```diff
@@ -76,6 +77,7 @@ public BackupTask(DocumentDatabase database, BackupParameters backupParameters,
             _forTestingPurposes = forTestingPurposes;
             _backupResult = GenerateBackupResult();
             TaskCancelToken = new OperationCancelToken(_database.DatabaseShutdown, CancellationToken.None);
+            MaxNumberOfFullBackupsInBackupHistory = database.ServerStore.Configuration.Backup.MaxNumberOfFullBackupsInBackupHistory;
```
this can be changed per database, so it should come from database configuration?

ppekrol commented on src/Raven.Server/Documents/PeriodicBackup/BackupTask.cs:
```diff
@@ -844,19 +846,6 @@ private void UploadToServer(string backupPath, string folderName, string fileNam
-                return;
-
-            // dismiss the previous operation
-            var id = $"{NotificationType.OperationChanged}/{_previousBackupStatus.LastOperationId.Value}";
```
for what this was for?

ppekrol commented on src/Raven.Server/ServerWide/Commands/PeriodicBackup/UpdatePeriodicBackupStatusCommand.cs:
```diff
@@ -29,9 +64,172 @@ protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawData
         {
             json[nameof(PeriodicBackupStatus)] = PeriodicBackupStatus.ToJson();
+            json[nameof(BackupHistoryEntries)] = new DynamicJsonArray(BackupHistoryEntries);
+            json[nameof(MaxNumberOfFullBackupsInBackupHistory)] = MaxNumberOfFullBackupsInBackupHistory;
```
why we need to persist that value in cluster?

ppekrol commented on src/Raven.Server/Web/System/BackupHistoryHandler.cs:
```diff
@@ -0,0 +1,64 @@
+
+namespace Raven.Server.Web.System;
+
+internal class BackupHistoryHandler : DatabaseRequestHandler
```
I think, due to the possibility of having excessive number of results, we need to consider paging here
What I think is:
1. In one endpoint return all full backups and the number of incrementals with paging
2. In second endpoint return all incrementals for given backup with paging

ppekrol commented on src/Raven.Server/Utils/BackupUtils.cs:
```diff
@@ -428,18 +430,23 @@ private static IdleDatabaseActivity GetNextIdleDatabaseActivity(NextIdleDatabase
         {
-            var command = new UpdatePeriodicBackupStatusCommand(databaseName, RaftIdGenerator.NewId()) { PeriodicBackupStatus = status };
+            PeriodicBackupStatus = status,
+            BackupHistoryEntries = documentDatabase.ConfigurationStorage.BackupHistoryStorage.RetractTemporaryStoredBackupHistoryEntries(),
```
if something goes wrong here, you already removed all temp backup entries, what you need to do is to retreve those and on successful command, you need to remove all the entries up to the last one you retrieved

ppekrol commented on test/Tests.Infrastructure/RavenTestBase.Backup.cs:
```diff
@@ -54,7 +54,7 @@ public async Task<long> RunBackupAsync(RavenServer server, long taskId, Document
                     var status = (await store.Maintenance.SendAsync(new GetOperationStateOperation(op))).Status;
                     return status;
-                }, opStatus, timeout: timeout ?? _reasonableTimeout);
+                }, opStatus, timeout: timeout ?? _reasonableTimeout * 10);
```
?

ppekrol commented on test/Tests.Infrastructure/RavenTestBase.Backup.cs:
```diff
@@ -225,7 +225,7 @@ public async Task RunBackupInClusterAsync(DocumentStore store, long taskId, bool
                     OperationStatus status = x.Status;
                     return status;
-                }, opStatus, timeout: timeout ?? _reasonableTimeout);
+                }, opStatus, timeout: timeout ?? _reasonableTimeout * 10);
```
?

ppekrol commented on src/Raven.Server/ServerWide/Commands/PeriodicBackup/UpdatePeriodicBackupStatusCommand.cs:
```diff
@@ -1,13 +1,48 @@
+        public List<BackupHistoryEntry> BackupHistoryEntries;
+        public int MaxNumberOfFullBackupsInBackupHistory;
+
+        public List<BackupHistoryEntry> CurrentAndTemporarySavedEntries
```
Convert to method

ppekrol commented on src/Raven.Server/ServerWide/Commands/PeriodicBackup/UpdatePeriodicBackupStatusCommand.cs:
```diff
@@ -29,9 +64,172 @@ protected override BlittableJsonReaderObject GetUpdatedValue(long index, RawData
+            CurrentAndTemporarySavedEntries.ForEach(
+                backupHistoryEntryToAdd =>
+                {
+                    var itemKey = backupHistoryEntryToAdd.GenerateItemKey();
```
IMO, the way this is persisted, makes in complex and a lot of data can be stored in one document, imagine a case when I have a full backup once a month and an incremental every 15min?

Each of those should be stored separately, yet should be identified by a key quicker for easy access, paging, removal, etc

ppekrol commented on src/Raven.Server/Utils/BackupHistoryWriter.cs:
```diff
@@ -0,0 +1,241 @@
+            _currentNodeTag = currentNodeTag;
+        }
+
+        public void GetBackupHistoryEntriesFromCluster(DocumentDatabase database, ServerStore serverStore, TransactionOperationContext context)
```
ClusterStateMachine should have a property Backups where all the methods for retrieving, etc are there

Writing should be as simple as creating in the endpoint writer and processing items retrieved from CSM using .ToJson() method

ppekrol commented:
Will be done as a part of backup rewrite in 6.1
